protocol/rpcconsumer/rpcconsumer_server.go

```diff
@@ -566,14 +651,27 @@ func (rpccs *RPCConsumerServer) sendRelayToProvider(
 		}
 	}
 
-	if reqBlock == spectypes.LATEST_BLOCK && protocolMessage.RelayPrivateData().SeenBlock != 0 {
-		// make optimizer select a provider that is likely to have the latest seen block
-		reqBlock = protocolMessage.RelayPrivateData().SeenBlock
+	addon := chainlib.GetAddon(protocolMessage)
+	reqBlock = rpccs.resolveRequestedBlockAndUpdateExtensionIfNeeded(reqBlock, protocolMessage.RelayPrivateData().SeenBlock, latestBlockHashRequested, earliestBlockHashRequested, addon, protocolMessage)
+	// check wether we need a new protocol message with the new earliest block hash requested
+	if earliestBlockHashRequested != spectypes.NOT_APPLICABLE {
```

Comment:
when you rad from the state archiveInfo, make sure to read it as the same time you read the protocolMessage, so you don;t have race windows, don't reach archiveInfo right now